### PR TITLE
Ft api addbooks 151690134

### PR DESCRIPTION
**Pull request by llabake**
#### What Does This PR Do?
This PR adds functionality to add books.
#### Description Of Task To Be Completed
Admin should be able to add books to the book object
#### Any Background Context You Want To Provide?
N/A
#### Screenshot(s)

image

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import express from 'express';
 import bodyParser from 'body-parser';
 import logger from 'morgan';
-// import routes from './server/routes';
+import routes from './server/routes';
 // import db from './server/models';
```
Comment (omokaro-faith) [inline app.js]: You would want to remove commented code

File: server/controllers/v1/bookController.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import Book from '../../models/book';
+
+export const addBook = (req, res) => {
+    try {
+        const book = new Book(req.body);
+        book.save();
+        return res.status(201).json({ message: `Book with ${book.title} has been added`, book })
+    } catch (error) {
+        console.log(error);
```
Comment (omokaro-faith) [inline server/controllers/v1/bookController.js]: Why are you logging error on this line?

File: server/helpers/modelHelpers.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  
+export const getObjectId = (objectType) => {
+    const dummyDataTypes = ['books'];
+    if (dummyDataTypes.indexOf(objectType) !== -1 ){
+      let id =  Object.keys(dummyData[objectType]).length + 1;
+      return id;
+    } else throw new Error (`dummyData type: ${objectType} not yet created`);
+}
+
+
```
Comment (omokaro-faith) [inline server/helpers/modelHelpers.js]: Extra line is needed here

File: server/models/book.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      for (let key in args) {
+        if (fields.indexOf(key) == -1) {
+          errors.push({ message: `invalid field: ${key} found in args` });
+        } else {
+          this[key] = args[key];        
+        }
+      }
+
+      if (errors.length > 0) {
+        // throw new Error (JSON.stringify(errors,null,4));
```
Comment (omokaro-faith) [inline server/models/book.js]: remove comment

File: server/models/book.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        throw errors;
+      }
+      this.deleted = 'false';
+      this.id = getObjectId('books');
+    }
+    
+    save () {
+      // console.log(`saving ${JSON.stringify(this)}`);
```
Comment (omokaro-faith) [inline server/models/book.js]: remove commented code

File: server/routes/indexRoute.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+const indexRoute = (app) => {
+    app.get('/', (req, res) => res.status(200).send({
+      message: 'Welcome to Hello Books',
+    })
+    );
+  };
+  
+export default indexRoute;
+
+
```
Comment (omokaro-faith) [inline server/routes/indexRoute.js]: extra line needed here

File: server/models/book.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      const allBooks = dummyData.books;
+      for (let id in allBooks) {
+        const book = allBooks[id];
+        delete book.deleted;
+        returnedBooks.push(book);
+      }
+      return returnedBooks;
+    }
+}
+
```
Comment (omokaro-faith) [inline server/models/book.js]: extra line needed here

File: server/models/book.js (hunk 4 above)
Comment (omokaro-faith) [reply server/models/book.js]: Line 26 can be modified to be just if(errors.length)

File: server/models/book.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        if (!args[field]) {
+          errors.push({ path: field, message: `${field} is required` });
+        } else if (typeof(args[field]) !== dataType[field]) {
+          errors.push({ path: field, message: `${field} must be a ${dataType[field]}` });
+        } 
+      });
+      
+      for (let key in args) {
+        if (fields.indexOf(key) == -1) {
+          errors.push({ message: `invalid field: ${key} found in args` });
```
Comment (omokaro-faith) [inline server/models/book.js]: Why you checking for if (fields.indexOf(Key) == -1) ?

File: app.js (hunk 1 above)
Comment (llabake) [reply app.js]: Changes have been made.

File: server/controllers/v1/bookController.js (hunk 2 above)
Comment (llabake) [reply server/controllers/v1/bookController.js]: It was an over sight. I have removed it in  my updated file

File: server/helpers/modelHelpers.js (hunk 3 above)
Comment (llabake) [reply server/helpers/modelHelpers.js]: Noted

File: server/models/book.js (hunk 8 above)
Comment (llabake) [reply server/models/book.js]: I am checking that the user does not supply a key that is not among my recognized fields.

File: server/models/book.js (hunk 4 above)
Comment (llabake) [reply server/models/book.js]: Noted

File: server/models/book.js (hunk 5 above)
Comment (llabake) [reply server/models/book.js]: Corrected

File: server/models/book.js (hunk 7 above)
Comment (llabake) [reply server/models/book.js]: Noted

File: server/routes/indexRoute.js (hunk 6 above)
Comment (llabake) [reply server/routes/indexRoute.js]: Corrected

**Pull request merged**